### DISCUSSION TITLE
Client Certificate Verification Failure on Server After Successful SSL/TLS Handshake

### DISCUSSION THREAD

**Discussion by aarti-programmer** [openssl/openssl#25149] [Q&A]
**Error with TLS Call Home connection using netopeer2-cli**

**Describe the issue:**
I am encountering an error when attempting to listen for a TLS Call Home connection using netopeer2-cli. Below is the command and the error log I receive:

**Command:**
listen --tls

**Error Log:**
cmd_listen: Waiting 60s for a TLS Call Home connection on port 4335...
nc ERROR: Communication socket unexpectedly closed.
cmd_listen: Receiving TLS Call Home on port 4335 failed.

Here are the relevant logs from the netopeer2-server:
[INF]: LN: Trying to connect via IPv4 to 172.17.167.137:4335.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Successfully connected to 172.17.167.137:4335 over IPv4.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Successfully connected to 172.17.167.137:4335 over IPv4.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Session 822 (user "root", CID 56) created.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: SR: Session 822 (user "root", CID 56) created.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Session 823 (user "root", CID 56) created.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: SR: Session 823 (user "root", CID 56) created.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Session 824 (user "root", CID 56) created.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: SR: Session 824 (user "root", CID 56) created.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Cert verify: depth 1.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Cert verify: depth 1.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Cert verify: subject: /C=in/ST=ut/L=ut/O=ut/OU=ru/CN=ut/emailAddress=ut@gmail.com.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Cert verify: issuer: /C=in/ST=ut/L=ut/O=ut/OU=ru/CN=ut/emailAddress=ut@gmail.com.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Cert verify CTN: cert fail, cert-to-name will continue on the next cert in chain.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Cert verify: depth 0.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Cert verify: subject: /C=in/ST=ut/L=ut/O=ut/OU=ru/CN=ut/emailAddress=ut@gmail.com.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Cert verify: issuer: /C=in/ST=ut/L=ut/O=ut/OU=ru/CN=ut/emailAddress=ut@gmail.com.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Cert-to-name unsuccessful, dropping the new client.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [ERR]: LN: **Client certificate error (application verification failure).**
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: [ERR]: LN: **SSL accept failed (certificate verify failed).**
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Cert verify: subject: /C=in/ST=ut/L=ut/O=ut/OU=ru/CN=ut/emailAddress=ut@gmail.com.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Cert verify: issuer: /C=in/ST=ut/L=ut/O=ut/OU=ru/CN=ut/emailAddress=ut@gmail.com.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Cert verify CTN: cert fail, cert-to-name will continue on the next cert in chain.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Cert verify: depth 0.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Cert verify: subject: /C=in/ST=ut/L=ut/O=ut/OU=ru/CN=ut/emailAddress=ut@gmail.com.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Cert verify: issuer: /C=in/ST=ut/L=ut/O=ut/OU=ru/CN=ut/emailAddress=ut@gmail.com.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: Cert-to-name unsuccessful, dropping the new client.
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: **Client certificate error (application verification failure).**
Aug 08 09:45:01 13266--SW--MCP7 netopeer2-server[41237]: **SSL accept failed (certificate verify failed).**
Aug 08 09:45:03 13266--SW--MCP7 netopeer2-server[41237]: Call Home client "default-client" endpoint "endpoint-tls" failed connection attempt limit 3 reached.
Aug 08 09:45:03 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Call Home client "default-client" endpoint "endpoint-tls" failed connection attempt limit 3 reached.
Aug 08 09:45:03 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Call Home client "default-client" endpoint "endpoint-tls" connecting...
Aug 08 09:45:03 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: Trying to connect via IPv4 to 172.17.167.137:4335.
Aug 08 09:45:03 13266--SW--MCP7 netopeer2-server[41237]: [INF]: LN: getsockopt() error (Connection refused).
Aug 08 09:45:03 13266--SW--MCP7 netopeer2-server[41237]: Call Home client "default-client" endpoint "endpoint-tls" connecting...
Aug 08 09:45:03 13266--SW--MCP7 netopeer2-server[41237]: Trying to connect via IPv4 to 172.17.167.137:4335.

**Steps to Reproduce:**
1. Start the netopeer2-server.
2. Run `listen --tls` command from netopeer2-cli.
3. Observe the error log.

**Expected Behavior:**
The TLS Call Home connection should be established successfully.

**Actual Behavior:**
The connection fails with the error: `SSL accept failed (certificate verify failed)`.
Any help resolving this issue would be greatly appreciated.

regard
aarti

**Comment by mattcaswell** [comment]
You are using TLS client authentication. It is clear from the error logs that the certificate being sent from the client to the server is failing to verify.

In your title you say "SSL_accept() failed in server with handshake error where SSL_connect() succeeds". This is actually normal behaviour for a client certificate verify failure where TLSv1.3 has been negotiated. `SSL_connect` and `SSL_accept` complete successfully when the TLS handshake completes successfully on each endpoint. The time this occurs is *different* on each endpoint. The handshake is considered complete when an endpoint has both sent and received a "Finished" message. 

If you look at figure 1 of section 2 in the RFC (https://datatracker.ietf.org/doc/html/rfc8446), you can see that this actually occurs on the client before it occurs on the server. The server sends its "Finished" message before the client. The client later sends its "Finished" message immediately after sending its certificate and certificate verify messages. At this point the client considers the handshake complete (and SSL_connect completes successfully). The server however has not yet processed the certificate/certificate verify/Finished messages because they are still in transit. When it does process those messages it spots the verification failure and aborts the handshake (and hence SSL_accept fails).

So the question is, why does the certificate fail to verify on the server? We get some clue from the message "Cert-to-name unsuccessful, dropping the new client." and the immediately preceding log lines. This suggests to me that it is failing to find the certificate or one of its roots in the trust store. But that error message does not come from OpenSSL, so it seems likely to me that the netopeer2-server is using some custom verify callback.

I would examine the certificate being sent from the client to the server and confirm that it looks valid, and that the server has been configured to trust it. However, I know nothing about netopeer2 so I can't provide any advice on how to do that. If you can't resolve this based on my info above, then it looks to me like you need to raise an issue with netopeer2. This doesn't look like an issue in OpenSSL itself.
